### PR TITLE
Support for GitLab CI/CD

### PR DESCRIPTION
**Pull request by ralphlange**
This branch adds support for GitLab's CI/CD service.

The examples and tests use `ubuntu:bionic` as Docker image, but any image from Dockerhub could be used.

### CODE REVIEW

**Comment by ralphlange** [comment]
@tboegi - I finished the GitLab support that I started before the summer as part of an attempt to collect open ends.
Please have a look - no hurry.

This branch is rebased on current `master` - it contains a few fixes that are not in a released version yet.

**Comment by tboegi** [comment]
@ralphlange 
A very quick test (on our local Gitlab server) failed like this.
(I need to dig deeper the next days)


$ perl --version
This is perl 5, version 26, subversion 1 (v5.26.1) built for x86_64-linux-gnu-thread-multi
(with 67 registered patches, see perl -V for more detail)
Copyright 1987-2017, Larry Wall
Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.
Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.
$ <unknown> --version
Traceback (most recent call last):
  File ".ci/cue.py", line 1133, in <module>
    main(sys.argv[1:])
  File ".ci/cue.py", line 1129, in main
    args.func(args)
  File ".ci/cue.py", line 970, in prepare
    sp.check_call([cc, '--version'])
  File "/usr/lib/python2.7/subprocess.py", line 185, in check_call
    retcode = call(*popenargs, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 172, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/usr/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory

**Comment by ralphlange** [comment]
Hm. The setting for the C/C++ compiler is empty.

The GitLab example files set CMP to either "gcc" or "clang". (So they work.) Should there be an implicit default if a user does not set the CMP variable at all?

**Comment by tboegi** [comment]
Sorry for the early alarm - CMP was not set up at all.
Improvements are ongoing

**Comment by tboegi** [comment]
@ralphlange: I got it basically working. 
Some small (unrelated to this PR) things are left to be fixed.
After that, I have a couple of small comments, please stay tuned ;-)

File: cue.py [hunk 1]
```diff
@@ -33,6 +33,7 @@ def detect_context():
         ci['service'] = 'gitlab'
         ci['os'] = 'linux'
         ci['platform'] = 'x64'
+        ci['sudo'] = []                    # No sudo in GitLab Docker containers
```
Comment (tboegi) [inline cue.py]: Would it make sense to install sudo ?
(Because many people need it anyway ?)

**Review by tboegi — COMMENTED** [review]
Looks good to me

File: gitlab/.gitlab-ci.yml.example-mini [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    - apt-get update -qq && apt-get install -y -qq build-essential git python curl p7zip-full
+    - python .ci/cue.py prepare
+  script:
+    - python .ci/cue.py build
+    - python .ci/cue.py test
+    - python .ci/cue.py test-results
+
+# Build on Linux using default gcc for Base branches 7.0 and 3.15
+
+gcc_7_0:
```
Comment (tboegi) [inline gitlab/.gitlab-ci.yml.example-mini:28]: This gcc_7_0 did confuse me: Are we really testing gcc 7.0 ?
Probaly yes, but we test gcc_epics_7_0, (or epics_7_0_gcc)

File: gitlab/README.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ - Can use any Docker image from Dockerhub (the examples use
+  `ubuntu:bionic`)
+ - Compile natively using different compilers (gcc, clang)
+ - Cross-compile for Windows 32bit and 64bit using MinGW and WINE
+ - Cross-compile for RTEMS 4.9 and 4.10 (Base >= 3.15)
+ - Built dependencies are cached (for faster builds).
+
+## How to Use these Scripts
+
+ 1. Get an account on [GitLab](https://gitlab.com/), create a project
```
Comment (tboegi) [inline gitlab/README.md:15]: Strictly speaking, some sites have a local installation of gitlab 
and then you don't need an account at gitlab.com

Comment (ralphlange) [reply gitlab/README.md:15]: Correct, but this applies to all CI services that we support: also GitHub, AppVeyor and Travis-CI offer "Enterprise" installations on company premises.

I will add a general sentence about this.

File: gitlab/.gitlab-ci.yml.example-mini (hunk 2 above)
Comment (ralphlange) [reply gitlab/.gitlab-ci.yml.example-mini:28]: Well, it's a job name. We could name it "Heinz" and "Horst".
Will improve the names.

File: gitlab/README.md (hunk 3 above)
Comment (ralphlange) [reply gitlab/README.md:15]: Remark added by way of 43e52b5f

File: gitlab/.gitlab-ci.yml.example-mini (hunk 2 above)
Comment (ralphlange) [reply gitlab/.gitlab-ci.yml.example-mini]: Improved by way of 984fa5e3

File: cue.py (hunk 1 above)
Comment (ralphlange) [reply cue.py]: It all depends on the people, what people are trying to do, and (most of all) which Docker image they configure for their build.

I'm against installing things solely based on what we assume people might need. The ci-scripts functionality works for the advertised scope without using sudo on GitLab, and adding `sudo` to the variable `APT` or to the install line in the `.build` job template is easy enough for any user who needs it for a container image that doesn't have it pre-installed.

**Comment by ralphlange** [comment]
Ok, thanks a lot for your test and review!

If the `sudo` thing really gets in your way, we can still add it later on.

**Pull request merged**